Link git outputs between tasks

shashwathi opened the pull request:
In this PR I have implemented outputs using PVC within lifetime of pipelinerun. Closed old PR #255 in favor of this. Discussion for design is in previous PR so please refer to that for more details.

Note:
- [x]  Replaces current e2e test for multiple tasks pipelinerun test with fan-in and fan-out pipeline. (More complex than existing e2e test and covers all use cases previous e2e test as well)
- [x] Added another example under `examples` to demonstrate git resource dependency
- [x] Support for targetPath is in pipeline task resource. e2e test includes this use case as well.

**What will not be included?**

Support for images
This PR only includes support for git resource dependency. Support for images includes some design of metadata and I would like to follow up with another PR for that.

/assign @ImJasonH
/assign @bobcatfish
/assign @pivotal-nader-ziada

pivotal-nader-ziada commented on pkg/apis/pipeline/v1alpha1/taskrun_types.go:
```diff
@@ -43,7 +43,19 @@ type TaskRunSpec struct {
+	// +optional
+	PVCName string `json:"pvcName,omitempty"`
+	// +optional
+	PreBuiltSteps []TaskBuildStep `json:"preBuildSteps,omitempty"`
```
Do we have to have these steps in the definition of the taskRun type as a field, can we not append them in the appropriate order while creating the taskRun

shashwathi replied:
> If not, can we somehow make it obvious we don't want users to provide them?

How about add validation in taskrun to check if owner reference is `pipelinerun Kind` then post build and pre build steps should be allowed else throw error.

vdemeester replied:
If we don't allow user to provide those field, it shouldn't be in the spec (should it ?), but I guess they are required for pipeline/pipelinerun to populate them… so… yeah most likely a validation error could work there.

Is there a use case for a user to provides those in addition to *normal* steps ?

bobcatfish replied:
> Is there a use case for a user to provides those in addition to normal steps ?

@vdemeester: @shashwathi and i discussed, and we were thinking that maybe a use case where a user wants to debug tasks that are part of a pipeline in isolation? it might be a bit of a stretch tho!

i think adding `paths` to the interface instead of the `builtSteps` makes it a bit clearer hopefully!

shashwathi commented on pkg/reconciler/v1alpha1/pipelinerun/resources/input_output_steps_test.go:
```diff
@@ -0,0 +1,233 @@
+		}},
+	}
+
+	resources.WrapSteps(taskRunSpec, pipelineResources, pt)
```
I had a similar thought about adding test for `WrapSteps` but I wanted to test the change to taskrun so I added a simple case to validate desired change happened.

shashwathi commented on test/crd_checks.go:
```diff
@@ -30,7 +30,7 @@ import (
 const (
 	interval = 1 * time.Second
-	timeout  = 5 * time.Minute
+	timeout  = 10 * time.Minute // TODO: timeout should be configurable via go test timeout
```
New pipelinerun e2e test includes 4 tasks and this takes upto 6min (on my gke) cluster. I wanted to add timeout to be on safer side so increased it upto 10min but it is very specific to that e2e test. So probably I will just update timeout for that test instead of all crds check timeout

pivotal-nader-ziada commented on test/helm_task_test.go:
```diff
@@ -255,8 +255,8 @@ func getHelmDeployPipeline(namespace string) *v1alpha1.Pipeline {
-					Name:       "workspace",
-					ProvidedBy: []string{createImageTaskName},
+					Name: "workspace",
+					//	ProvidedBy: []string{createImageTaskName}, //Yet to be implemented
```
Do you mean you are going to implement this as part of this same PR?

shashwathi commented on test/pipelinerun_test.go:
```diff
@@ -213,23 +219,177 @@ func getHelloWorldPipelineWithSingularTask(suffix int, namespace string) *v1alph
+					Name:    "read-something-else",
+					Image:   "ubuntu",
+					Command: []string{"cat"},
+					Args:    []string{"/workspace/readingspace/something-else"},
```
> what do you think if we had the tasks themselves assert? i.e. have the step fail if the expected text doesnt exist, e.g. using grep
I like the idea. I will update the tasks to assert contents.

> im a bit confused by what each task and each step in each task is doing - maybe we could use the names of the steps to make it a bit more clear

I tried to name tasks with "stuff" and "something" because the tasks are actually writing the same words. I could update them to pattern you mentioned.

shashwathi commented on pkg/apis/pipeline/v1alpha1/taskrun_types.go:
```diff
@@ -43,7 +43,19 @@ type TaskRunSpec struct {
+	// +optional
+	PreBuiltSteps []TaskBuildStep `json:"preBuildSteps,omitempty"`
+	// +optional
+	PostBuiltSteps []TaskBuildStep `json:"postBuiltSteps,omitempty"`
```
I leaned toward `Built ` because thats the end object we are creating now. I like the name `BeforeSteps` for pre build steps, for post build steps we could go with `AfterSteps`.

shashwathi commented on pkg/reconciler/v1alpha1/pipelinerun/resources/input_output_steps.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+var (
+	pvcDir = "/pvc"
```
I had thought of not creating `/pvc` directory at all. Then I realized that pvc mounted with `/` path will dump all task named folders so putting directories under some named path would be better. I will update the docs

shashwathi commented on pkg/apis/pipeline/v1alpha1/resource_types.go:
```diff
@@ -118,6 +119,8 @@ type PipelineResource struct {
 	Name        string              `json:"name"`
 	ResourceRef PipelineResourceRef `json:"resourceRef"`
+	// +optional
+	SourcePaths []string `json:"sourcePaths"`
```
I do not think I am using that. I will go ahead and delete that field

shashwathi commented on pkg/reconciler/v1alpha1/taskrun/resources/pre_post_build_step.go:
```diff
@@ -0,0 +1,124 @@
+				}
+				newSteps = append(newSteps, []corev1.Container{{
+					Name:         fmt.Sprintf("source-mkdir-%s", source.Name),
+					Image:        "busybox",
```
Can I follow up with another PR to fix image?

shashwathi commented:
@pivotal-nader-ziada @bobcatfish @vdemeester 

I updated the design in recent change. Please take a look at it and leave comments. Thanks to @bobcatfish for slack brainstorming.

- Updated e2e test to assert contents
- No more use of steps(post, pre). Using `taskrun.resource.paths` to accomplish passing information to taskrun.
- Updated docs.
- Updated commit msg to include issue number and more details.

bobcatfish commented on test/crd_checks.go:
```diff
@@ -73,12 +73,12 @@ func WaitForPodState(c *clients, name string, namespace string, inState func(r *
 // error or timeout. desc will be used to name the metric that is emitted to
 // track how long it took for name to get into the state checked by inState.
-func WaitForPipelineRunState(c *clients, name string, inState func(r *v1alpha1.PipelineRun) (bool, error), desc string) error {
+func WaitForPipelineRunState(c *clients, name string, polltimeout time.Duration, inState func(r *v1alpha1.PipelineRun) (bool, error), desc string) error {
```
:+1:

bobcatfish commented on test/pipelinerun_test.go:
```diff
@@ -213,23 +225,172 @@ func getHelloWorldPipelineWithSingularTask(suffix int, namespace string) *v1alph
+					Name:    "read-from-task-0",
+					Image:   "ubuntu",
+					Command: []string{"bash"},
+					Args:    []string{"-c", "[[ stuff == $(cat /workspace/stuff) ]]"},
```
niiiiiice :D

bobcatfish commented:
/lgtm

In case you want to make some tweaks to the docs - altho it's _unlikely_ a user would want to use the `paths` directly, I think it would be nice to try to show how they could do that:

/hold

And also I'm still wondering about fan-out and fan-in, seems like we can end up with tasks running in parallel changing the same data?

shashwathi commented:
> In case you want to make some tweaks to the docs - altho it's unlikely a user would want to use the paths directly, I think it would be nice to try to show how they could do that:

I will take another shot at docs tomorrow. 👍

shashwathi commented:
@bobcatfish : I have updated the docs to include examples. Please review that and let me know if you have any feedback.

bobcatfish commented on docs/Concepts.md:
```diff
@@ -169,6 +199,66 @@ Creating a `TaskRun` will invoke a [Task](#task), running all of the steps until
 
+`TaskRun` definition includes `inputs`, `outputs` for `Task` referred in spec.
+
+Input resource includes name and reference to pipeline resource and optionally `paths`. `paths` will be used by `TaskRun` as the resource's new source paths i.e., copy the resource from specified list of paths. `TaskRun` expects the folder and contents to be already present in specified paths. `paths` feature could be used to provide extra files or altered version of existing resource before exec ...
```
ah great, I get it now! thanks for the example above and the extra detail in the docs :D

bobcatfish commented:
/lgtm
/meow space

shashwathi commented:
@bobcatfish : I need `/approve` label too.

bobcatfish commented:
whoops, just took it for granted that @shashwathi had approver rights already, she's one of the top contributors!!

/approve

bobcatfish commented:
/hold cancel
